### PR TITLE
[Dataset] Adapter to convert a dataset for link prediction task

### PR DESCRIPTION
**Pull request by VoVAllen**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/data/adapter.py [hunk 1]
```diff
@@ -108,3 +114,181 @@ def __getitem__(self, idx):
...
+                 split_ratio=[0.8, 0.1, 0.1],
+                 neg_ratio=3,
+                 add_self_loop=True,
+                 **kwargs):
+        self.g = dataset[0]
+        self.dataset = dataset
+        self.split_ratio = split_ratio
+        self.neg_ratio = neg_ratio
+        self.add_self_loop = add_self_loop
+        super().__init__(dataset.name + '-as-edgepred', hash_key=(neg_ratio, split_ratio, add_self_loop), **kwargs)
```
Comment (jermainewang) [inline python/dgl/data/adapter.py:188]: If `hash_key` is designed for proper re-loading. Please help change `AsNodePred` as well. Make sure you've tested them properly.

File: tests/compute/test_data.py [hunk 2]
```diff
@@ -1067,6 +1067,19 @@ def test_as_nodepred2():
...
+
+@unittest.skipIf(F._default_context_str == 'gpu', reason="Datasets don't need to be tested on GPU.")
+def test_as_edgepred():
+    # create
+    ds = data.AsEdgePredDataset(data.CoraGraphDataset(), split_ratio=[0.8, 0.1, 0.1])
+    # Cora has 10556 edges, 10% test edges can be 1057
+    assert ds.get_test_edges()[0][0].shape[0] == 1057
+    # read from cache
+    ds = data.AsEdgePredDataset(data.CoraGraphDataset(), split_ratio=[0.8, 0.1, 0.1])
+    assert ds.get_test_edges()[0][0].shape[0] == 1057
```
Comment (jermainewang) [inline tests/compute/test_data.py:1080]: Please add a test for invalid cache.

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/data/adapter.py [hunk 3]
```diff
@@ -108,3 +114,181 @@ def __getitem__(self, idx):
 
     def __len__(self):
         return 1
+
+
+
+def negative_sample(g, num_samples):
```
[the remark this answers is not in the document]
Comment (VoVAllen) [reply python/dgl/data/adapter.py:120]: It's not guaranteed. Can be less than the num_samples, which is consistent to DGL's negative sample. Here using numpy is just to ensure reproducibility with seed provided.

File: python/dgl/data/adapter.py [hunk 4]
```diff
@@ -108,3 +114,181 @@ def __getitem__(self, idx):
...
+                graph, self.neg_ratio*(n_val+n_test))
+            neg_n_val, neg_n_test = self.neg_ratio * n_val, self.neg_ratio * n_test
+            neg_val_src, neg_val_dst = neg_src[:neg_n_val], neg_dst[:neg_n_val]
+            neg_test_src, neg_test_dst = neg_src[neg_n_val:], neg_dst[neg_n_val:]
+            self.val_edges = (F.tensor(src[val_pos_idx]), F.tensor(dst[val_pos_idx])
+                              ), (F.tensor(neg_val_src), F.tensor(neg_val_dst))
+            self.test_edges = (F.tensor(src[test_pos_idx]),
+                               F.tensor(dst[test_pos_idx])), (F.tensor(neg_test_src), F.tensor(neg_test_dst))
+            self.train_graph = create_dgl_graph(
+                (src[train_pos_idx], dst[train_pos_idx]), num_nodes=self.num_nodes)
```
[the remark this answers is not in the document]
Comment (VoVAllen) [reply python/dgl/data/adapter.py:228]: But for generated edges, the val_edges and test_edges should be masked out in the training stgae right?

File: python/dgl/data/adapter.py [hunk 5]
```diff
@@ -75,9 +79,11 @@ def process(self):
         if 'label' not in self.g.nodes[self.target_ntype].data:
             raise ValueError("Missing node labels. Make sure labels are stored "
                              "under name 'label'.")
-        if self.verbose:
-            print('Generating train/val/test masks...')
-        utils.add_nodepred_split(self, self.split_ratio, self.target_ntype)
+        if any(s not in self.g.nodes[self.target_ntype].data for s in ["train_mask", "val_mask", "test_mask"]):
```
[the remark this answers is not in the document]
Comment (VoVAllen) [reply python/dgl/data/adapter.py]: fixed

File: python/dgl/data/adapter.py [hunk 6]
```diff
@@ -60,48 +65,233 @@ class AsNodePredDataset(DGLDataset):
...
+    differs from the cached one, it will re-process the dataset properly.
+
+    Parameters
+    ----------
+    dataset : DGLDataset
+        The dataset to be converted.
+    split_ratio : (float, float, float), optional
+        Split ratios for training, validation and test sets. Must sum to one.
+    neg_ratio : int, optional
+        Indicate how much negative samples to be sampled
```
Comment (jermainewang) [inline python/dgl/data/adapter.py:165]: How many negative samples(edges) to generate per positive sample(edge).

File: python/dgl/data/adapter.py [hunk 7]
```diff
@@ -2,11 +2,16 @@
...
+from ..convert import graph as create_dgl_graph
+from ..sampling.negative import _calc_redundancy
 from .dgl_dataset import DGLDataset
 from . import utils
 
-__all__ = ['AsNodePredDataset']
+__all__ = ['AsNodePredDataset', 'AsEdgePredDataset']
+
 
 class AsNodePredDataset(DGLDataset):
```
Comment (jermainewang) [inline python/dgl/data/adapter.py:16]: Add a `feat_size` property to this class is not a bad idea.

File: python/dgl/data/adapter.py [hunk 8]
```diff
@@ -60,48 +65,233 @@ class AsNodePredDataset(DGLDataset):
...
+    """Repurpose a dataset for edge prediction task.
+
+    The created dataset will include data needed for link prediction.
+    Currently only support homogeneous graph. 
+    It will keep only the first graph in the provided dataset and
+    generate train/val/test edges according to the given split ratio,
+    and the correspondent negative edges based on the neg_ratio. The generated
+    edges will be cached to disk for fast re-loading. If the provided split ratio
+    differs from the cached one, it will re-process the dataset properly.
+
```
Comment (jermainewang) [inline python/dgl/data/adapter.py:157]: Only support dataset of homogeneous graphs.

Fix numpy random seed to generate deterministic train/val/test split.

**Review by jermainewang — APPROVED** [review]
Remember to:
* Rename `AsEdgePredDataset` to `AsLinkPredDataset`.
* Add docstring for `val_edges` and `test_edges`.

Otherwise, LGTM

**Pull request merged**